Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +25,62 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+   if (current_ < 0)
+   {
+       return true;
+   }
+   else
+   {
+       return false;
+   }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return timeWhenChargedOrDepleted_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
     Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+
+    if (checkStart_ == true)
+    {
+        totalAmpHoursUsed_ = (initialStateOfChargePercent_ / 100) * BATTERY_AMP_HOUR_CAPACITY;
+    }
+
+    else if (checkStart_ == false)
+    {
+        timeDiff_ = (double)(time_.msecsTo(batteryData.time) * MSECONDS_TO_HOURS_CONVERSION);
+        double currentAverage = (batteryData.current + current_) / 2;
+        totalAmpHoursUsed_ += currentAverage * timeDiff_;
+    }
+
+    time_ = batteryData.time;
+
+    current_ = batteryData.current;
+
+    checkStart_ = false;
```

Comment:
This only needs to be set if checkStart_ was true, so we can place that in the if above.